lib/pbench/cli/agent/commands/tools/kill.py

```diff
@@ -132,17 +133,14 @@ def gen_host_names(tm_dir: pathlib.Path) -> str:
     of remote hosts.
     """
     run_dir = tm_dir.parent
-    tool_group_dirs = list(run_dir.glob(f"{ToolGroup.TOOL_GROUP_PREFIX}-*"))
-    if not tool_group_dirs:
+    tool_groups = list(gen_tool_groups(run_dir))
```

Comment:
One should not care about the other, right?  I don't want to influence the interface by this one use case.

The other use case was elusive to find (see https://github.com/distributed-system-analysis/pbench/blob/main/lib/pbench/cli/agent/commands/tools/base.py#L28) and that is a property.  It might be better to remove `gen_tool_groups` and fix the property to do it correctly (not hard-coding the prefix).

And thanks for opening this can of worms.

Reply:
Oh, wow, and, 

```
            [p.name.split("tools-v1-")[1] for p in self.pbench_run.glob("tools-v1-*")]
```

Using the entire tools/version string as the split separator?

And, yes; another use case, but also immediately list-ifying the results, so not a demonstration of a "real" generator.

While I agree that the method API shouldn't depend on a single caller pattern, we also don't want a design that essentially forces a common caller "mitigation" pattern everyone now "needs to know about" like list-ifying. And here, for example, we sort in addition: should the "generator" be sorting the output at the source?

Reply:
Can we defer all this to PR #3038?